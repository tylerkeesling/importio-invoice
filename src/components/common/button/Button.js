import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import styles from './button.module.css';

const propTypes = {
  text: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  btnType: PropTypes.string.isRequired,
};

const defaultProps = {
  disabled: false,
};

// create constant for button types
// usage would be <Button.Add {...props} />
const TYPES = {
  ADD: 'btn-add',
  DELETE: 'btn-delete',
};

const BaseButton = ({ text, onClick, disabled, btnType }) => {
  const btnClass = classnames(styles.btn, styles[`${btnType}`]); // concatenating different button styles

  return (
    <button
      type="button" // cant be a prop
      disabled={disabled}
      onClick={onClick}
      className={btnClass}
    >
      {text}
    </button>
  );
};

BaseButton.propTypes = propTypes;
BaseButton.defaultProps = defaultProps;

// exporting both ADD and DELETE buttons
export const Add = props => <BaseButton {...props} btnType={TYPES.ADD} />;

export const Delete = props => <BaseButton {...props} btnType={TYPES.DELETE} />;
