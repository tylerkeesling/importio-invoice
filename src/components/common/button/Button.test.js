import React from 'react';
import { mount } from 'enzyme';
import * as Button from './Button';

const mockFn = jest.fn();

const setup = propOverrides => {
  const props = Object.assign(
    {
      text: 'This is my label',
      onClick: mockFn,
    },
    propOverrides
  );

  const wrapper = mount(<Button.Add {...props} />);

  return {
    props,
    wrapper,
  };
};

describe('<Button />', () => {
  it('matches snapshot', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call function from onClick callback when clicked', () => {
    const { wrapper } = setup();
    wrapper.simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });

  it('should have the correct label/text', () => {
    const { wrapper } = setup();
    expect(wrapper.find('BaseButton').text()).toEqual('This is my label');
  });
});
