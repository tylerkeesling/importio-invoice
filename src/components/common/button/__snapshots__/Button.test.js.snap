// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches snapshot 1`] = `
ReactWrapper {
  Symbol(enzyme.__unrendered__): <Add
    onClick={[MockFunction]}
    text="This is my label"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "onClick": [MockFunction],
      "text": "This is my label",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "btnType": "btn-add",
        "disabled": false,
        "onClick": [MockFunction],
        "text": "This is my label",
      },
      "ref": null,
      "rendered": Object {
        "instance": <button
          class="btn btn-add"
          type="button"
        >
          This is my label
        </button>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "This is my label",
          "className": "btn btn-add",
          "disabled": false,
          "onClick": [MockFunction],
          "type": "button",
        },
        "ref": null,
        "rendered": Array [
          "This is my label",
        ],
        "type": "button",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "onClick": [MockFunction],
        "text": "This is my label",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "btnType": "btn-add",
          "disabled": false,
          "onClick": [MockFunction],
          "text": "This is my label",
        },
        "ref": null,
        "rendered": Object {
          "instance": <button
            class="btn btn-add"
            type="button"
          >
            This is my label
          </button>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "This is my label",
            "className": "btn btn-add",
            "disabled": false,
            "onClick": [MockFunction],
            "type": "button",
          },
          "ref": null,
          "rendered": Array [
            "This is my label",
          ],
          "type": "button",
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
